src/coreclr/vm/comdelegate.cpp

```diff
@@ -791,6 +942,50 @@ LoaderHeap *DelegateEEClass::GetStubHeap()
     return GetInvokeMethod()->GetLoaderAllocator()->GetStubHeap();
 }
 
+#if defined(TARGET_RISCV64)
+static Stub* CreateILDelegateShuffleThunk(MethodDesc* pDelegateMD, MethodDesc* pTargetMD)
+{
+    // TODO: Stack growing is a rare case but if thunk duplication turns out to be an issue, optimize by cooking
+    // a stripped signature similar to GenerateShuffleArray and use call target from DelegateObject::GetMethodPtrAux()
+    _ASSERTE(pDelegateMD->SizeOfArgStack() < pTargetMD->SizeOfArgStack()); // IL thunk handle stack growing only
+    SigTypeContext typeContext(pDelegateMD);
+    MetaSig delegateSig(pDelegateMD);
+    INDEBUG(MetaSig targetSig(pTargetMD));
+    LOG((LF_STUBS, LL_INFO1000000,
+        "CreateILDelegateShuffleThunk %s.%s (%i args) -> %s.%s (%i args + %i this)\n",
+        pDelegateMD->GetMethodTable()->GetDebugClassName(), pDelegateMD->GetName(), delegateSig.NumFixedArgs(),
+        pTargetMD->GetMethodTable()->GetDebugClassName(), pTargetMD->GetName(), targetSig.NumFixedArgs(), !!targetSig.HasThis()
+    ));
+    _ASSERTE(delegateSig.NumFixedArgs() == targetSig.NumFixedArgs() + (targetSig.HasThis() ? 1 : 0));
+    _ASSERTE(delegateSig.GetReturnType() == targetSig.GetReturnType());
+    _ASSERTE(delegateSig.HasThis());
+
+    ILStubLinkerFlags flags = ILSTUB_LINKER_FLAG_STUB_HAS_THIS;
+    ILStubLinker stubLinker(pDelegateMD->GetModule(), pDelegateMD->GetSignature(), &typeContext, pDelegateMD, flags);
+    ILCodeStream *pCode = stubLinker.NewCodeStream(ILStubLinker::kDispatch);
+
+    for (unsigned i = 0; i < delegateSig.NumFixedArgs(); ++i)
+        pCode->EmitLDARG(i);
+
+    pCode->EmitCALL(pCode->GetToken(pTargetMD), delegateSig.NumFixedArgs(), delegateSig.IsReturnTypeVoid() ? 0 : 1);
+    pCode->EmitRET();
+
+    PCCOR_SIGNATURE pSig;
+    DWORD cbSig;
+    pDelegateMD->GetSig(&pSig, &cbSig);
+    PTR_Module pLoaderModule = pDelegateMD->GetLoaderModule();
+    MethodDesc* pStubMD = ILStubCache::CreateAndLinkNewILStubMethodDesc(
+        pDelegateMD->GetLoaderAllocator(),
+        pDelegateMD->GetMethodTable(),
+        ILSTUB_DELEGATE_SHUFFLE_THUNK,
+        pDelegateMD->GetModule(),
+        pSig, cbSig,
+        &typeContext,
+        &stubLinker);
+
+    return Stub::NewStub(JitILStub(pStubMD));
```

Comment:
Where is the Stub allocated by `Stub::NewStub` going to be freed?

Reply:
Could leak, how should the lifetime be managed for IL shuffling thunks?

Reply:
Avoid calling `Stub::NewStub` for the IL stub case. It is not necessary.

Reply:
Right, done.